src/node/genesis_gen.h

```diff
@@ -351,6 +351,12 @@ namespace ccf
         return false;
       }
 
+      if (active_service->status == ServiceStatus::OPEN)
+      {
+        // If the service is already open, return with no effect
+        return true;
+      }
```

Comment:
As a general principle for making these proposals idempotent I think we should prefer "do it again", rather than "early out" (ie - adding `ServiceState::OPEN` as one of the acceptable states in the `if` below). I see 3 main arguments for this:
- Simplifying audit. If a proposal says `do_foo` and I expect that to write `did: FOO` to the ledger, then that write should always be present in a transaction which passes a `do_foo` transaction. I shouldn't need to track the invariants against existing state to reproduce the service's write logic, I should wherever possible be able to reason about these writes as atoms.
- Supporting re-application across code versions. If we change this proposal to _also_ write `magic_flag: 42` to the KV, on line 370, then it looks like that magic flag is set in every service which has opened. But services which migrate from the old code version to the new won't have this, and may have no way to set it. Rather than handling this manually, or with a separate proposal, the neatest solution seems to be to allow the members to "open the service again", and trust that always applies the latest invariants for an open service. This is harder where it might involve merging with current state, but that can be handled case-by-case.
- Consistent tracking of write conflicts. Early outs remove some of our interaction with the KV, so that the retry semantics of the original and re-applied proposal are different. I don't have a full story for the transaction-logic bug this introduces, but there's enough of a smell that I expect its there (or at least - I know there's something unknown there).

Reply:
We discussed this offline, and there's some major arguments in favour of the "idempotence means some proposals are successful nops" approach:
- Smaller ledger. If you're deploying through some abstract infrastructure that ends up spamming your proposals, its safe because they're idempotent, and better if they don't produce spam in the ledger (maybe they can even become complete nops, not writing to the proposals table if they instantly pass and have no effect, we shouldn't rule this possibility out).
- Moving the complexity elsewhere. If we re-apply changes, then every hook that expects to be called once needs to be robust to this.
- Precondition checks need to be precise in either case. Most of the second case above boils down to being able to distinguish "have I done this before", and what that means across multiple code versions with different effects. Avoiding writes may be safer than always writing, since the bad outcome is "can't do the thing I wanted" rather than "produced inconsistent state", until the constitution is updated/fixed.
- The example as written _would_ produce a conflict, because A read `service_status` to determine whether it had already been applied. A similar hole could occur if the reads and writes were not aligned - by reading `X` I decide not to write `Y`, and thus miss a conflict with `Y`. But this devolves to the "need to check your preconditions precisely" noted above.

In summary we'd prefer the early-out approach for current use cases, and if we find a service that would prefer re-application we'll need to provide it in open form, without ruling out the optimisations possible from the nop approach.

Reply:
@eddyashton Thanks for the write-up! I've linked it in the parent issue: https://github.com/microsoft/CCF/issues/1791#issuecomment-804820088